airbyte-cdk/java/airbyte-cdk/db-sources/src/testFixtures/java/io/airbyte/cdk/testutils/ContainerFactory.java

```diff
@@ -22,97 +23,73 @@
  * ContainerFactory is the companion interface to {@link TestDatabase} for providing it with
  * suitable testcontainer instances.
  */
-public interface ContainerFactory<C extends JdbcDatabaseContainer<?>> {
+public abstract class ContainerFactory<C extends JdbcDatabaseContainer<?>> {
+
+  static private final Logger LOGGER = LoggerFactory.getLogger(ContainerFactory.class);
+
+  private record ContainerKey(Class<? extends ContainerFactory> clazz, DockerImageName imageName, List<String> methods) {};
+
+  private record ContainerOrError(GenericContainer<?> container, RuntimeException e) {};
+
+  private static final ConcurrentMap<ContainerKey, ContainerOrError> SHARED_CONTAINERS = new ConcurrentHashMap<>();
 
   /**
    * Creates a new, unshared testcontainer instance. This usually wraps the default constructor for
    * the testcontainer type.
    */
-  C createNewContainer(DockerImageName imageName);
-
-  /**
-   * Returns the class object of the testcontainer.
-   */
-  Class<?> getContainerClass();
+  protected abstract C createNewContainer(DockerImageName imageName);
```

Comment:
I like that this move to an abstract class allows this to be `protected` and prevents clients from calling this method directly.